File: Creational/FactoryMethod/Java/Structure/Creator.java
/**
 * Creator abstract class - định nghĩa factory method
 * và có thể chứa business logic khác
 */
public abstract class Creator {
    
    /**
     * Factory Method - phương thức abstract để tạo Product
     * Các subclass sẽ override để tạo concrete product cụ thể
     */
    public abstract Product factoryMethod();
    
    /**
     * Business logic sử dụng factory method
     * Lưu ý: method này gọi factoryMethod() nhưng không biết
     * concrete product nào sẽ được tạo ra
     */
    public void someOperation() {
        System.out.println("Creator: Bắt đầu thực hiện nghiệp vụ...");
        
        // Gọi factory method để tạo product
        Product product = factoryMethod();
        
        System.out.println("Creator: Đang làm việc với " + product.getName());
        product.operation();
        
        System.out.println("Creator: Hoàn thành thực hiện nghiệp vụ.");
    }
}

File: Creational/FactoryMethod/Python/Structure/creator.py
"""
Creator class - declares the factory method that returns Product objects
"""
from abc import ABC, abstractmethod
from product import Product


class Creator(ABC):
    """
    Abstract Creator class - declares the factory method that returns a Product object
    """
    
    @abstractmethod
    def factory_method(self) -> Product:
        """
        Factory method that concrete creators must implement
        Note: Despite its name, the Creator's primary responsibility is not creating products.
        Usually, it contains some core business logic that relies on Product objects, 
        returned by the factory method.
        
        Returns:
            Product: A concrete product instance
        """
        pass
    
    def some_operation(self) -> str:
        """
        Some operation that uses the product created by the factory method
        Note: The Creator's code does not depend on concrete product classes
        
        Returns:
            str: Result of the operation using the created product
        """
        # Create a product using the factory method
        product = self.factory_method()
        
        # Use the product
        result = f"Creator: Working with {product.operation()}"
        
        return result
